xdsl/dialects/irdl.py

```diff
@@ -0,0 +1,389 @@
+"""Definition of the IRDL dialect."""
+
+from __future__ import annotations
+from typing import Annotated, Iterable, Sequence
+from xdsl.dialects.builtin import StringAttr, SymbolRefAttr
+from xdsl.ir import (
+    Attribute,
+    Block,
+    Dialect,
+    OpResult,
+    ParametrizedAttribute,
+    Region,
+    SSAValue,
+    TypeAttribute,
+)
+from xdsl.irdl import (
+    IRDLOperation,
+    OpAttr,
+    SingleBlockRegion,
+    VarOperand,
+    irdl_attr_definition,
+    irdl_op_definition,
+)
+from xdsl.parser import Parser
+from xdsl.printer import Printer
+from xdsl.traits import HasParent
+
+################################################################################
+# Dialect, Operation, and Attribute definitions                                #
+################################################################################
+
+
+@irdl_attr_definition
+class AttributeType(ParametrizedAttribute, TypeAttribute):
```

Comment:
How do you get by with only an AttributeType and no TypeType? That feels like spoilering the plot that they actually are the same...

This makes the OperandsOp and ResultsOp look a bit weird, as it now wants AttributeTypes, whereas MLIR convention seems to indicate a TypeType there.

Reply:
So the reason is that to factor the code, and make things more concise and simple, I put everything in `Attribute`.
If I have a type, it is directly put in a `TypeAttr` in MLIR.
This has a small runtime cost in MLIR, but we don't care that much about runtime yet